### ISSUE TITLE
federation: verify and add tests for replicaset preferences annotations

### ISSUE THREAD

**Issue by irfanurrehman** [kubernetes/federation#74]
<a href="https://github.com/nikhiljindal"><img src="https://avatars0.githubusercontent.com/u/10199099?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [nikhiljindal](https://github.com/nikhiljindal)**
_Thursday Sep 22, 2016 at 20:08 GMT_
_Originally opened as https://github.com/kubernetes/kubernetes/issues/33312_

----

Found while writing the federation replicaset guide:

Steps to repro:
- Register a cluster C1 for federation.
- Create a federation replicaset with X replicas.
- Observe that a replicaset is created in the cluster C1 with X replicas.
- Now register another cluster C2 for federation.

Expected:
- The replicaset is also created in cluster C2 with X/2 replicas. Replica set in cluster C1 is scaled down from X to X/2 replicas.

Actual:
- Replicaset not created in cluster C2. Replicaset in cluster C1 continues to have X replicas.

cc @kubernetes/sig-cluster-federation

**Comment by irfanurrehman** [comment]
<a href="https://github.com/nikhiljindal"><img src="https://avatars0.githubusercontent.com/u/10199099?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [nikhiljindal](https://github.com/nikhiljindal)**
_Thursday Sep 22, 2016 at 20:09 GMT_

----

I see now that we do not have any e2e test for this case (register a cluster after a replicaset is created)

**Comment by irfanurrehman** [comment]
<a href="https://github.com/nikhiljindal"><img src="https://avatars0.githubusercontent.com/u/10199099?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [nikhiljindal](https://github.com/nikhiljindal)**
_Thursday Sep 22, 2016 at 20:10 GMT_

----

cc @kshafiee @jianhuiz

**Comment by irfanurrehman** [comment]
<a href="https://github.com/nikhiljindal"><img src="https://avatars0.githubusercontent.com/u/10199099?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [nikhiljindal](https://github.com/nikhiljindal)**
_Thursday Sep 22, 2016 at 21:14 GMT_

----

Reading the description of [rebalance](https://github.com/kubernetes/kubernetes/blob/cf4ea938c882d5f86852994807b18959458033d5/federation/apis/federation/types.go#L121) field in FederatedReplicaSetPreferences, it looks like rebalancing is expected not to happen unless rebalance is set to true.

I tried setting rebalance = true in the annotation on the federated replicaset before creating it and it still didnt rebalance.

**Comment by irfanurrehman** [comment]
<a href="https://github.com/jianhuiz"><img src="https://avatars1.githubusercontent.com/u/1171596?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jianhuiz](https://github.com/jianhuiz)**
_Thursday Sep 22, 2016 at 22:08 GMT_

----

you need to specify it like this

```
  annotations:
    federation.kubernetes.io/replica-set-preferences: >
      {
        "rebalance": true,
        "clusters": {
          "*": {
            "weight": 1
          }
        }
      }
```

**Comment by irfanurrehman** [comment]
<a href="https://github.com/nikhiljindal"><img src="https://avatars0.githubusercontent.com/u/10199099?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [nikhiljindal](https://github.com/nikhiljindal)**
_Friday Sep 23, 2016 at 17:27 GMT_

----

Yes I realised I wasnt doing it right the first time.
It seemed as if all was fine, but then I saw controller manager logs and saw errors in parsing my annotation. Ideally my create request should fail if the annotation is incorrect.
I got it right finally (no annotation parsing errors), but my replicaset still didnt get rebalanced when I added a cluster.
We should really add an e2e test for this.

This is not a release blocker since the annotation is alpha.

**Comment by irfanurrehman** [comment]
<a href="https://github.com/jianhuiz"><img src="https://avatars1.githubusercontent.com/u/1171596?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jianhuiz](https://github.com/jianhuiz)**
_Friday Sep 23, 2016 at 21:39 GMT_

----

@nikhiljindal I wasn't able to reproduce this, could you upload the controller logs and the replicaset file you use?

**Comment by irfanurrehman** [comment]
<a href="https://github.com/nikhiljindal"><img src="https://avatars0.githubusercontent.com/u/10199099?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [nikhiljindal](https://github.com/nikhiljindal)**
_Friday Sep 30, 2016 at 21:11 GMT_

----

The annotation I had tried I think was:

```
annotations:
    federation.kubernetes.io/replica-set-preferences: >
      {
        "rebalance": true,
        }
      }
```

We should add a test case for that.
Can you or @mwielgus take that up?

**Comment by irfanurrehman** [comment]
<a href="https://github.com/jianhuiz"><img src="https://avatars1.githubusercontent.com/u/1171596?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jianhuiz](https://github.com/jianhuiz)**
_Monday Oct 03, 2016 at 15:45 GMT_

----

not specifying weight would implies 0 for all clusters which results in no replicaset would be created in the underlying clusters.

**Comment by irfanurrehman** [comment]
<a href="https://github.com/mwielgus"><img src="https://avatars2.githubusercontent.com/u/11994812?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [mwielgus](https://github.com/mwielgus)**
_Monday Oct 03, 2016 at 16:35 GMT_

----

Yep, i can take it.

**Comment by irfanurrehman** [comment]
<a href="https://github.com/nikhiljindal"><img src="https://avatars0.githubusercontent.com/u/10199099?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [nikhiljindal](https://github.com/nikhiljindal)**
_Tuesday Nov 08, 2016 at 19:29 GMT_

----

Another replicaset issue reported on slack: we do not set the "ready" field in replicaset status: https://kubernetes.slack.com/archives/sig-federation/p1478597937000904

**Comment by irfanurrehman** [comment]
<a href="https://github.com/jianhuiz"><img src="https://avatars1.githubusercontent.com/u/1171596?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jianhuiz](https://github.com/jianhuiz)**
_Thursday Nov 10, 2016 at 04:49 GMT_

----

and AvailableReplicas, too, which are introduced later to fed replicaset. I'll add them tomorrow.
#29481 #32771

**Comment by irfanurrehman** [comment]
<a href="https://github.com/csbell"><img src="https://avatars1.githubusercontent.com/u/23373339?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [csbell](https://github.com/csbell)**
_Friday Jan 06, 2017 at 01:54 GMT_

----

@mwielgus, I can take this one.

**Comment by irfanurrehman** [comment]
<a href="https://github.com/henriquetruta"><img src="https://avatars2.githubusercontent.com/u/2501482?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [henriquetruta](https://github.com/henriquetruta)**
_Friday May 12, 2017 at 19:47 GMT_

----

@nikhiljindal @jianhuiz didn't work for me. Used the annotation as you suggested and the replicas weren't rebalanced this is the yaml I've used
```
apiVersion: extensions/v1beta1
kind: ReplicaSet
metadata:
  name: nginx-rs-bal1
spec:
  replicas: 6 
  template:
    metadata:
      labels:
        app: nginx1
      annotations:
        federation.kubernetes.io/replica-set-preferences: >
          {
            "rebalance": true,
            "clusters": {
              "*": {
                "weight": 1
              }
            }
          }
    spec:
      containers:
      - name: nginx
        image: nginx:1.7.9
        ports:
        - containerPort: 80
```

**Comment by irfanurrehman** [comment]
<a href="https://github.com/nikhiljindal"><img src="https://avatars0.githubusercontent.com/u/10199099?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [nikhiljindal](https://github.com/nikhiljindal)**
_Monday Jun 05, 2017 at 17:35 GMT_

----

Updated the issue title to be generic to add tests for replica set preferences.
IIUC, there are no tests for these annotations at all.

@henriquetruta is taking a first cut at https://github.com/kubernetes/kubernetes/pull/45840

**Comment by irfanurrehman** [comment]
<a href="https://github.com/kustuntas"><img src="https://avatars2.githubusercontent.com/u/8240288?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kustuntas](https://github.com/kustuntas)**
_Friday Jun 09, 2017 at 18:20 GMT_

----

@nikhiljindal Has the rebalancing issue been fixed yet?

**Comment by irfanurrehman** [comment]
<a href="https://github.com/nikhiljindal"><img src="https://avatars0.githubusercontent.com/u/10199099?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [nikhiljindal](https://github.com/nikhiljindal)**
_Saturday Jun 10, 2017 at 01:15 GMT_

----

@kustuntas I havent tried it again but it doesnt seem to be based on updates on this bug.

cc @perotinus

**Comment by irfanurrehman** [comment]
<a href="https://github.com/quinton-hoole"><img src="https://avatars0.githubusercontent.com/u/10390785?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [quinton-hoole](https://github.com/quinton-hoole)**
_Friday Sep 08, 2017 at 03:01 GMT_

----

Unassigning @mwielgus, who I'm pretty sure is not working on this.  RE-assign if needed.
Sounds like we have both a missing test (addressed in #45840), and possibly a bug.
We'll need to resolve this to GA federation.

**Comment by irfanurrehman** [comment]
cc @nikhiljindal
